docs/async-await.md

```diff
@@ -0,0 +1,232 @@
+
+```swift
+    public var trailers: HTTPHeaders? { async throws }
+```
+
+However we will need to make sure that the user has consumed the body stream completely before, calling the trailers, because otherwise we might run into a situation from which we can not progress forward:
+
+```swift
+do {
+  var request = HTTPClientRequest(url: "https://swift.org/")
+  let response = try await httpClient.execute(request, deadline: .now() + .seconds(3))
+  
+  var trailers = try await response.trailers // can not move forward since body must be consumed before.
+} catch {
+  print(error)
+}
+```
+
+In such a case we can either throw an error or crash.
+
+### New invocation
+
+The new way to invoke a request shall look like this:
+
+```swift
+extension HTTPClient {
+  func execute(_ request: HTTPClientRequest, deadline: NIODeadline) async throws -> HTTPClientResponse
+}
+```
+
+Simple usage example:
+
+```swift
+var request = HTTPClientRequest(url: "https://swift.org")
+request.method = .POST
+request.headers = [
+  "content-type": "text/plain; charset=UTF-8"
+  "x-my-fancy-header": "super-awesome"
+]
+request.body = .bytes("Hello world!".utf8)
+
+var response = try await client.execute(request, deadline: .now() + .seconds(5))
+
+switch response.status {
+case .ok:
+  let body = try await response.body.collect(maxBytes: 1024 * 1024)
+default:
+  throw MyUnexpectedHTTPStatusError
+}
+```
+
+Stream upload and download example using the new `FileHandle` api:
+
+```swift
+let readHandle = FileHandle(forReadingFrom: URL(string: "file:///tmp/readfile")!))
+let writeHandle = FileHandle(forWritingTo: URL(string: "file:///tmp/writefile")!))
+
+var request = HTTPClientRequest(url: "https://swift.org/echo")
+request.method = .POST
+request.headers = [
+  "content-type": "text/plain; charset=UTF-8"
+  "x-my-fancy-header": "super-awesome"
+]
+request.body = .stream(readHandle.bytes)
+
+var response = try await client.execute(request, deadline: .now() + .seconds(5))
+
+switch response.status {
+case .ok:
+  Task {
+    var streamIterator = response.body.makeAsyncIterator()
+    writeHandle.writeabilityHandler = { handle in
+      switch try await streamIterator.next() {
+      case .some(let buffer):
+        handle.write(buffer.readData(buffer.readableBytes)!)
+      case .none:
+        handle.close()
+      }
+    }
+  }
+default:
+  throw MyUnexpectedHTTPStatusError
+}
+```
+
+- **Why do we have a deadline in the function signature?** 
+    Task deadlines are not part of the Swift 5.5 release. However we think that they are an important tool to not overload the http client accidentally. For this reason we will not default them.
+- **What happened to the Logger?** We will use Task locals to propagate the logger metadata. @slashmo and @ktoso are currently working on this.
+- **How does cancellation work?** Cancellation works by cancelling the surrounding task:
+
+        ```swift
+        let task = Task {
```

Comment:
Swift already has special syntax to create a child task through `async let`. What you propose would create a detached Task. I think this would probably make it too easy and users would create detached Tasks just because there is a method for it, even if they would actually want a child Task through `async let`.